### PR TITLE
Unit tests for PBFT

### PR DESCRIPTION
**Pull request by bridger-herman**
Scope: Commits https://github.com/bitwiseio/sawtooth-pbft/pull/5/commits/86c0d887675c44f505b52bdfdf4b19bcde374b36 through https://github.com/bitwiseio/sawtooth-pbft/pull/5/commits/3d86ee8a3d650716532721fa85221a384117f4bf (8 commits)

Changes introduced:
+ Add unit tests for:
  + `timing`
  + `state`
  + `message_log`
  + `node`
+ Add materials for running unit tests in Docker

### CODE REVIEW

File: src/timing.rs [hunk 1]
```diff
@@ -89,3 +89,59 @@ impl Timeout {
...
+                panic!(
+                    "Value is not within tolerance ({:?} - {:?} > {:?})",
+                    $val1, $val2, $tol
+                );
+            }
+        };
+    }
+
+    #[test]
+    fn ticker() {
```
Comment (agunde406) [inline src/timing.rs]: Please add comments to the test describing what is being tested

File: Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,7 @@ protobuf = "2"
 clap = "2.31"
 log = "0.4"
 log4rs = "0.8"
+rust-crypto = "0.2"
```
Comment (agunde406) [inline Cargo.toml]: If this is only used for testing it can be moved to a [dev-dependencies] section of the Cargo.toml file.

File: src/main.rs [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 #[macro_use]
 extern crate clap;
+extern crate crypto;
```
Comment (agunde406) [inline src/main.rs]: If only used in tests add #[cfg(test)] above the extern crate

File: src/state.rs [hunk 4]
```diff
@@ -266,3 +266,82 @@ impl PbftState {
...
+    #[test]
+    fn no_fault_tolerance() {
+        let config = mock_config(1);
+        let caught = ::std::panic::catch_unwind(|| {
+            PbftState::new(0, &config);
+        }).is_err();
+        assert!(caught);
+    }
+
+    #[test]
```
Comment (agunde406) [inline src/state.rs]: add comments describing what is being tested to all tests

File: src/message_log.rs [hunk 5]
```diff
@@ -437,3 +437,184 @@ fn num_unique_signers(msg_info_list: &[&PbftMessageInfo]) -> u64 {
...
+        let peer_id_map: HashMap<u64, PeerId> = cfg
+            .peers
+            .clone()
+            .into_iter()
+            .map(|(peer_id, node_id)| (node_id, peer_id))
+            .collect();
+        peer_id_map[&which].clone()
+    }
+
+    #[test]
```
Comment (agunde406) [inline src/message_log.rs]: add comments describing what is being tested to all tests

File: src/node.rs [hunk 6]
```diff
@@ -937,3 +938,368 @@ fn get_block_by_id(service: &mut Box<Service>, block_id: &BlockId) -> Option<Blo
...
+        assert_eq!(node1.state.phase, PbftPhase::NotStarted);
+        assert_eq!(
+            node1.state.working_block,
+            WorkingBlockOption::NoWorkingBlock
+        );
+        assert_eq!(node1.state.seq_num, 0);
+    }
+
+    #[test]
+    fn block_valid() {
```
Comment (agunde406) [inline src/node.rs]: add comments describing what is being tested to all tests

File: tests/sawtooth-pbft-test.dockerfile [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 Intel Corporation
+# Copyright 2018 Bitwise IO, Inc.
```
Comment (agunde406) [inline tests/sawtooth-pbft-test.dockerfile]: If this file started off as a copy of an intel file it should include both copyrights so it looks like:

Copyright 2018 Intel Corporation
Copyright 2018 Bitwise IO, Inc.

**Review by agunde406 — CHANGES_REQUESTED** [review]

**Comment by bridger-herman** [comment]
@agunde406 Updated to address your thoughts. Added comments to all tests describing their purpose

**Review by agunde406 — APPROVED** [review]

**Review by jsmitchell — APPROVED** [review]

**Pull request merged**